spatialmath/geometry_utils.go

```diff
@@ -50,30 +52,72 @@ func SegmentDistanceToSegment(ap1, ap2, bp1, bp2 r3.Vector) float64 {
 }
 
 // ClosestPointsSegmentSegment will return the points at which two line segments are closest to one another.
-func ClosestPointsSegmentSegment(ap1, ap2, bp1, bp2 r3.Vector) (r3.Vector, r3.Vector) {
-	// vectors between line endpoints:
-	v0 := bp1.Sub(ap1)
-	v1 := bp2.Sub(ap1)
-	v2 := bp1.Sub(ap2)
-	v3 := bp2.Sub(ap2)
-
-	// squared distances:
-	d0 := v0.Norm2()
-	d1 := v1.Norm2()
-	d2 := v2.Norm2()
-	d3 := v3.Norm2()
-
-	// select best potential endpoint on capsule A:
-	bestA := ap1
-	if d2 < d0 || d2 < d1 || d3 < d0 || d3 < d1 {
-		bestA = ap2
-	}
-	// select point on capsule B line segment nearest to best potential endpoint on A capsule:
-	bestB := ClosestPointSegmentPoint(bp1, bp2, bestA)
-
-	// now do the same for capsule A segment:
-	bestA = ClosestPointSegmentPoint(ap1, ap2, bestB)
-	return bestA, bestB
+func ClosestPointsSegmentSegment(a1, a2, b1, b2 r3.Vector) (r3.Vector, r3.Vector) {
+	// Proceed according to this math.stackexchange answer: https://math.stackexchange.com/a/2812513
+	// The stack solution has a sign error, which we correct
+	// These (ugly) variable names were chosen to match those in the stack exchange answer, for easy reference
+
+	d3121 := b1.Sub(a1).Dot(a2.Sub(a1))
+	d4321 := b2.Sub(b1).Dot(a2.Sub(a1))
+	d4331 := b2.Sub(b1).Dot(b1.Sub(a1))
+	r1pow2 := a2.Sub(a1).Norm2()
+	r2pow2 := b2.Sub(b1).Norm2()
+
+	denom := r1pow2*r2pow2 - d4321*d4321
+	// If denom is 0, the segments are parallel, and we can jump to the endpt case below
+	// If denom is not 0, we finish our algebra
+	if denom != 0 {
+		// Find the closest points on the lines each segment define
+		// If s or t lie outside of [0,1], the closest points on the lines are NOT on the finite segments
+		// Is overflow a concern?
+		s := (d3121*r2pow2 - d4331*d4321) / denom
+		t := (d3121*d4321 - d4331*r1pow2) / denom
+		if 0 <= s && s <= 1 && 0 <= t && t <= 1 {
+			bestSeg1Pt := a1.Add(a2.Sub(a1).Mul(s))
+			bestSeg2Pt := b1.Add(b2.Sub(b1).Mul(t))
+			return bestSeg1Pt, bestSeg2Pt
+		}
+	}
+
+	// If we're here, the lines are either parallel or their closest points lie off at least one of the segments
+	// It suffices to just check each segment against the other segment's endpoints
+	bestSeg1Pt := a1
+	bestSeg2Pt := ClosestPointSegmentPoint(b1, b2, a1)
+	bestDist := bestSeg1Pt.Sub(bestSeg2Pt).Norm2() // actually squared distance, but doesn't matter
+	if bestDist == 0 {
```

Comment:
Same as above; we will essentially never get a true 0 here, but something below some float epsilon will be "good enough".